merchantapi/src/main/java/fi/hel/verkkokauppa/configuration/api/merchant/MerchantController.java

```diff
@@ -47,20 +48,24 @@ public MerchantController(MerchantService merchantService, Environment env, Encr
     @PostMapping(value = "/merchant/upsert", produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<MerchantDto> upsertMerchant(@RequestBody @Valid MerchantDto merchantDto) {
         try {
+            MerchantDto returnDto;
 
             // Checking if the merchant already exists. If it does, it will update the merchant.
             if (merchantDto.getMerchantId() != null) {
-                // Returning a response with a status code of 201 (CREATED) and the body of the response is the
-                // merchantDto.
-                return ResponseEntity.status(HttpStatus.CREATED).body(
-                        merchantService.update(merchantDto)
-                );
+                returnDto = merchantService.update(merchantDto);
+            } else {
+                returnDto = merchantService.save(merchantDto);
             }
 
-            // Returning a response with a status code of 201 (CREATED) and the body of the response is the merchantDto.
-            return ResponseEntity.status(HttpStatus.CREATED).body(
-                    merchantService.save(merchantDto)
-            );
+            String merchantPaytrailMerchantId = returnDto.getMerchantPaytrailMerchantId();
+            if (merchantPaytrailMerchantId != null) {
+                PaytrailMerchantMappingDto paytrailMerchantMappingDto = merchantService.getPaytrailMerchantMappingByMerchantPaytrailMerchantIdAndNamespace(merchantPaytrailMerchantId, returnDto.getNamespace());
```

Comment:
Should you add some kind of validation to prevent secret update if return value is null etc?

Reply:
What this function does if merchant exists and have merchantSecret set to it, should we prevent updating of secret value if it set to non null value and return value of mapping returns null/empty result?

Reply:
Why would there be a mapping with null secret in it? New merchants must supply merchantPaytrailMerchantId, so what's the point? Existing merchants don't have merchantPytrailMerchantId and thus no mapping is fetched for them.

Reply:
I dont see any null value checks for returning paytrail secret value before updating it?

``` 
if

```